### DISCUSSION TITLE
How are offsets calculated?

### DISCUSSION THREAD

**Discussion by aaronshenhao** [steveruizok/perfect-freehand#20] [Q&A] [answered]
Hi,

I'm quite amazed how smooth the lines are! I'm quite curious as to how you came up with the formula for calculating the offsets when adding points ([link](https://github.com/steveruizok/perfect-freehand/blob/8c5a9907e7df2a44c5cdf46a93678a2f7e65e104/src/index.ts#L261)):  `const offset = vec.mul(vec.per(vec.lrp(nextVector, vector, dpr)), radius)`.

I noticed that it used to be `const offset = vec.mul(vec.per(vector), radius)` in an older [commit](https://github.com/steveruizok/perfect-freehand/commit/201a545955d653f6681e301028ba0b9bb238d9fa). There's also currently a similar looking formula for finding the offset when drawing caps at sharp corners ([link](https://github.com/steveruizok/perfect-freehand/blob/8c5a9907e7df2a44c5cdf46a93678a2f7e65e104/src/index.ts#L236)): `const offset = vec.mul(vec.per(prevVector), radius)`.

I tried drawing a few diagrams but it doesn't seem obvious to me how it works (why linearly interpolate the dot product between the two unit vectors?). Also, when drawing caps at a point, why is the offset `vec.mul(vec.per(prevVector), radius)` rather than `vec.mul(vec.per(vector), radius)`?

Thanks!

**Comment by steveruizok** [comment, marked as answer]
Ah the many weird corners of this algorithm. The answer is generally “lots of trial and error” mixed with “trying everything and seeing what worked”.

The first question has to do with pressure: the more pressure, the greater the radius, and so the greater the risk (iirc) of strange corners when a corner is not quite enough to create a cap. You can try the using the old offset code and drawing some upper case “M”s, and see if you can spot the difference.

The second question is easier to answer specifically: if we find a sharp corner, then the current vector is thrown out; we need to start the cap using the previous vector (which would have pointed to the current point).
